lxd/identities.go

```diff
@@ -46,72 +44,72 @@ func updateCertificateCache(d *Daemon) {
 		return
 	}
 
-	for i, dbCert := range dbCerts {
-		_, found := newCerts[dbCert.Type]
-		if !found {
-			newCerts[dbCert.Type] = make(map[string]x509.Certificate)
+	identityCacheEntries := make([]identity.CacheEntry, 0, len(identities))
+	var localServerCerts []cluster.Certificate
+	for _, id := range identities {
+		cacheEntry := identity.CacheEntry{
+			Identifier:           id.Identifier,
+			AuthenticationMethod: string(id.AuthMethod),
+			IdentityType:         string(id.Type),
+			Projects:             projects[id.ID],
 		}
 
-		certBlock, _ := pem.Decode([]byte(dbCert.Certificate))
-		if certBlock == nil {
-			logger.Warn("Failed decoding certificate", logger.Ctx{"name": dbCert.Name, "err": err})
-			continue
-		}
+		if cacheEntry.AuthenticationMethod == api.AuthenticationMethodTLS {
+			cert, err := id.X509()
+			if err != nil {
+				logger.Warn("Failed to extract x509 certificate from TLS identity metadata", logger.Ctx{"error": err})
+				continue
+			}
 
-		cert, err := x509.ParseCertificate(certBlock.Bytes)
-		if err != nil {
-			logger.Warn("Failed parsing certificate", logger.Ctx{"name": dbCert.Name, "err": err})
-			continue
+			cacheEntry.Certificate = cert
 		}
 
-		newCerts[dbCert.Type][shared.CertFingerprint(cert)] = *cert
-
-		if dbCert.Restricted {
-			newProjects[shared.CertFingerprint(cert)] = certs[i].Projects
-		}
+		identityCacheEntries = append(identityCacheEntries, cacheEntry)
 
 		// Add server certs to list of certificates to store in local database to allow cluster restart.
-		if dbCert.Type == certificate.TypeServer {
-			localCerts = append(localCerts, dbCert)
+		if id.Type == api.IdentityTypeCertificateServer {
+			cert, err := id.ToCertificate()
+			if err != nil {
+				logger.Warn("Failed to convert TLS identity to server certificate", logger.Ctx{"error": err})
+			}
+
+			localServerCerts = append(localServerCerts, *cert)
 		}
 	}
 
 	// Write out the server certs to the local database to allow the cluster to restart.
-	err = s.DB.Node.Transaction(context.TODO(), func(ctx context.Context, tx *db.NodeTx) error {
-		return tx.ReplaceCertificates(localCerts)
+	err = s.DB.Node.Transaction(d.shutdownCtx, func(ctx context.Context, tx *db.NodeTx) error {
```

Comment:
Maybe we should use context.Background() here specifically in the write part so we dont read the certs and then potentially abort writing them out, what do you think?

Reply:
Hmmm... I'm not sure. If LXD shuts down the database, do we wait for all transactions to complete first? If not, passing in this context would at least make sure that the transaction is aborted cleanly.

Reply:
yeah the transaction would be aborted cleanly, if thats ok we can leave as is.

Reply:
I think it's ok. It's worth noting that this is replacing certificates in the local database on every cluster db heartbeat. We can probably be cleverer about this and use the cache to determine if they have changed or not and only write if they have. For now it doesn't seem to be a performance concern though.